components/brave_today/browser/locales_helper.cc

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_today/browser/locales_helper.h"
+
+#include <string>
+#include <vector>
+
+#include "base/containers/flat_map.h"
+#include "base/containers/flat_set.h"
+#include "base/feature_list.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_today/browser/publishers_controller.h"
+#include "brave/components/brave_today/browser/urls.h"
+#include "brave/components/brave_today/common/brave_news.mojom-forward.h"
+#include "brave/components/brave_today/common/brave_news.mojom-shared.h"
+#include "brave/components/brave_today/common/brave_news.mojom.h"
+#include "brave/components/brave_today/common/features.h"
+#include "third_party/abseil-cpp/absl/types/optional.h"
+
+namespace brave_news {
+namespace {
+bool HasAnyLocale(const base::flat_set<std::string>& locales,
+                  const mojom::Publisher* publisher) {
+  for (const auto& locale : publisher->locales) {
+    if (locales.contains(locale))
+      return true;
+  }
+  return false;
+}
+
+absl::optional<std::string> GetBestMissingLocale(
+    const base::flat_set<std::string>& locales,
+    const std::vector<mojom::Publisher*> publishers) {
+  base::flat_map<std::string, uint32_t> missing_locale_counts;
+  for (auto const* publisher : publishers) {
+    // If this publisher is already covered by the list of locales we don't want
+    // its locales to skew the list of what's missing.
+    if (HasAnyLocale(locales, publisher))
+      continue;
+
+    for (const auto& locale : publisher->locales)
+      missing_locale_counts[locale]++;
+  }
+
+  absl::optional<std::string> most_common;
+  uint32_t max_occurences = 0;
+  for (const auto& it : missing_locale_counts) {
+    if (it.second > max_occurences) {
+      most_common = it.first;
+      max_occurences = it.second;
+    }
+  }
+  return most_common;
+}
+
+}  // namespace
+
+base::flat_set<std::string> GetPublisherLocales(const Publishers& publishers) {
+  base::flat_set<std::string> result;
+  for (const auto& it : publishers) {
+    for (const auto& locale : it.second->locales)
+      result.insert(locale);
```

Comment:
nit: I think this could be

```suggestion
    result.insert(it.second->locales.begin(), it.second->locales.end());
```